### DISCUSSION TITLE
How works not-found.js with router dynamic?

### DISCUSSION THREAD

**Discussion by denisso** [vercel/next.js#45927] [Help]
When 404 is triggered, the page is completely reloaded, and I get a 404 page which is the default.

I have questions:
- Why does the page reload when 404 is triggered?
- Why doesn't mine work not-found.js ?

Link to [deploy](https://6951jo-3000.preview.csb.app/) and [code repository](https://github.com/denisso/next13-app-exp/tree/router-dynamic ) Next.js version canary.10 and canary.14

app file structure
```
app
│   globals.css
│   layout.module.scss
│   layout.tsx
│   not-found.tsx
│   page.tsx
│
├───dynamicparams-false
│   │   layout.tsx
│   │   not-found.tsx
│   │   page.tsx
│   │
│   └───[slug]
│           page.tsx
│
└───dynamicparams-true
    │   layout.tsx
    │   not-found.tsx
    │   page.tsx
    │
    └───[slug]
            page.tsx
```

file: dynamicparams-false \ [slug] \ page.tsx

```tsx
import { Page as PageComponent, TProps } from "@/components/page";

export async function generateStaticParams() {
  return [{slug: "1"}, {slug: "2"}];
}

export default function Page(props: TProps) {
  return <PageComponent {...props} />;
}
export const dynamicParams = false;
```

To reproduce the problem, go to the Dynamic Params false section and select one of the gray links

**Comment by icyJoseph** [comment]
Hi,

The problem is that, as of now, to show the `not-found` component one must call the `notFound` function from `next/navigation`:

```js
import { notFound } from "next/navigation";

import { Page as PageComponent, TProps } from "@/components/page";

export async function generateStaticParams() {
  return [{ slug: "1" }, { slug: "2" }];
}

export default function Page(props: TProps) {
  if (Number(props.params.slug) > 2) {
    notFound();
  }

  return <PageComponent {...props} />;
}

export const dynamicParams = true;
```

Since you have `export const dynamicParams = false` in:

- /dynamicparams-false/layout.tsx
- /dynamicparams-false/page.tsx

I think you also have to change/remove those, and that should do the trick.

From the docs:

> Invoking the notFound() function throws a NEXT_NOT_FOUND error and terminates rendering of the route segment in which it was thrown. Specifying a [not-found file](https://beta.nextjs.org/docs/api-reference/file-conventions/not-found) allows you to gracefully handle such errors by rendering a Not Found UI within the segment.


I understand your request though, and I am not sure there's a way to do that right now. Beta stage things 😅

**Reply by denisso** [reply]
Thanks for the answer, yes it work! But i try to find problems and write about them. 

In the "pages" file 404.js triggered always when the page is not found, or when getStaticProps returns {props: {NotFound: true}} . And it's convenient, making forced calls to notFound() doesn't seem convenient to me.

**Reply by icyJoseph** [reply]
Yeah I agree and most likely so does the Next team. On the appDir feedback thread they talked about this, but it's such a long thread now that I can't find the post.

On the docs they also talk about, if I remember correctly, implementing a 404-like solution for appDir.

The real problem, for me, is not calling to `notFound` (I think that's very convenient and just as clear as `notFound: true`) but tis that dynamicPaths set to `true`, then the request still has to hit the server and be rendered, and it is not just automatic ISR-like. That's not very efficient I think. 

In ISR you return `notFound: true`, and that's saved for the next requests, until revalidation time.
